### PR TITLE
Move private class calls to after the class definition

### PR DESCRIPTION
**Pull request by GoodForOneFare**
This builds on the `hoist_private_class_vars` PR.  Private/static calls within a class are now moved outside/below the class definition.

Sample conversions:

**`tax_code_column.coffee`** (the original intent of this mixin seems wonky, but it still translates)

``` coffeescript
class Shopify.TaxCodeColumn extends Shopify.AutocompleteV2.SingleSelectRemoteSourceAutocomplete
  Shopify.Mixin(@prototype, Shopify.SpreadsheetColumn::)
  # SNIP class code.
```

``` js
Shopify.TaxCodeColumn = class TaxCodeColumn extends Shopify.AutocompleteV2.SingleSelectRemoteSourceAutocomplete {
  // SNIP class code.
};

Shopify.Mixin(TaxCodeColumn.prototype, Shopify.SpreadsheetColumn.prototype);
```

---

**`theme_editor/data/block.coffee`**

``` coffeescript
class ThemeEditor.Block
  _.extend(@prototype, EventMixin)
  # SNIP class code.
```

``` js
ThemeEditor.Block = class Block {
  // SNIP class code.
}
_.extend(Block.prototype, EventMixin);
```

---

**`theme_editor/pickers/color_picker.coffee`**

``` coffeescript
class ThemeEditor.Pickers.Color extends ThemeEditor.Picker
  @claimSettingType('color')
  # SNIP class code.
```

``` js
ThemeEditor.Pickers.Color = class Color extends ThemeEditor.Picker {
  // SNIP class code.
}

Color.claimSettingType("color");
```

/cc @lemonmade, @fandy, @justinthec

### CODE REVIEW

File: src/parser.js [hunk 1]
```diff
@@ -642,6 +632,13 @@ function mapStatement(node, meta) {
   return b.expressionStatement(mapExpression(node, meta));
 }
 
+function getClassName(classNode) {
+  const nameProperties = classNode.variable.properties;
+  const lastName = nameProperties[nameProperties.length - 1].name.value;
+
+  return lastName;
```
Comment (lemonmade) [inline src/parser.js]: Unnecessary temporary variable

File: src/parser.js [hunk 2]
```diff
@@ -652,13 +649,41 @@ function mapClassPrivateVariables(classNode, meta) {
...
+
+  if (privateCalls.length === 0) {
+    return [];
+  }
+
+  if (!classNode.variable) {
+    throw new Error('Anonymous classes cannot contain private calls');
+  }
+
+  jsc(privateCalls).find(jsc.MemberExpression, {
```
Comment (lemonmade) [inline src/parser.js:672]: Can you look for a `ThisExpression` instead?

**Comment by lemonmade** [comment]
Great job 👍 Similar comments to last time. On our side, can we maybe add a note of warning to `esify` about the fact that we are moving around private class members/ calls? I don't think we will run into any issues for our code but it would be good to tell people what we are doing with things that don't have native ES2015 equivalents (again, would be easy to find examples where this could break).

**Comment by GoodForOneFare** [comment]
Raised https://github.com/Shopify/javascript/issues/164 to warn users about private stuff being shuffled around (great suggestion, thank you!)

File: src/parser.js (hunk 2 above)
Comment (GoodForOneFare) [inline src/parser.js:672]: Unfortunately, no.  `decaf` is generating vanilla JS objects for these arguments, so looser checks are necessary.

_mocha 2016-06-13 10-26-30

File: src/parser.js (hunk 1 above)
Comment (GoodForOneFare) [inline src/parser.js]: Taken care of.  ty!

**Comment by GoodForOneFare** [comment]
Raising a new PR that targets master.

**Comment by GoodForOneFare** [comment]
Superseded by https://github.com/Shopify/decaf/pull/12.

**Pull request closed without merging**